### DISCUSSION TITLE
Add and manage combined styles to apply in components

### DISCUSSION THREAD

**Discussion by TomFreudenberg** [tailwindlabs/tailwindcss#12579] [Help]
Hi,

I am using tailwindcss for a while now and really like it, even when I couldn't see this good workflow from the beginning on. I am not sure if I have understand everything quite yet but I will continue to learn ;-)

---

Currently I am trying to figure out how to solve some project request and how to manage that easy with tailwindcss.

The example comes from working with ShadcnUI.

Let's take the Dialog element as an example.

In case for seeing and getting updates, I try not to modify the original codes at all but to store them for monitoring in repository:

/shadcn/ui/dialog.tsx

```node
...
export { Dialog, DialogPortal, DialogOverlay, DialogClose, DialogTrigger, DialogContent, DialogHeader, DialogFooter, DialogTitle, DialogDescription };
```

For every of the components I created the `allow to update` components in my project like:

/mydesign/ui/dialogs.tsx

```node
...
import { Dialog, DialogPortal, DialogOverlay, DialogClose, DialogTrigger, DialogContent, DialogHeader, DialogFooter, DialogTitle, DialogDescription } from '@/shadcn/ui/dialog';

export { Dialog, DialogPortal, DialogOverlay, DialogClose, DialogTrigger, DialogContent, DialogHeader, DialogFooter, DialogTitle, DialogDescription };
```

and use only them inside the propject, like:

/app/page.tsx

```node
import { Dialog } from '@/mydesign/ui/dialog';
```

The reason is overwriting and maintaining currently.

So I don't like the default for dialog with the `ring` class attached and replace the component like:

/mydesign/ui/dialogs.tsx

```node
...
import { Dialog, DialogPortal, DialogOverlay, DialogClose, DialogTrigger, DialogContent, DialogHeader, DialogFooter, DialogTitle, DialogDescription } from '@/shadcn/ui/dialog';

const MyDesignDialogContent = React.forwardRef<
  React.ElementRef<typeof DialogContent>,
  React.ComponentPropsWithoutRef<typeof DialogContent>
>(({ className, children, ...props }, ref) => (
  <DialogPortal>
    <DialogOverlay />
    <DialogContent
      ref={ref}
      className={cn(
        'fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg',
        className
      )}
      {...props}
    >
      {children}
      <DialogClose className="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground">
        <X className="h-4 w-4" />
        <span className="sr-only">Close</span>
      </DialogClose>
    </DialogContent>
  </DialogPortal>
));

export { Dialog, DialogPortal, DialogOverlay, DialogClose, DialogTrigger, MyDesignDialogContent as DialogContent, DialogHeader, DialogFooter, DialogTitle, DialogDescription };
```

I can now change the the parts necessary to change like:

```node
// change right-8 top-8 rounded-xl
//
      <DialogClose className="absolute right-8 top-8 rounded-xl opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground">
```

So far so good - that works.

But now an every project I have to make changes on a number of element field, allways fitting the project.

I would like to add something like a "special" class string to make it more configuarble:

like:

```node
// change right-8 top-8 rounded-xl
//
      <DialogClose className={cn("absolute right-4 top-4 rounded-sm ...", my_special_classes)} 
```

so that must be a string holding the necessary changes like:

const my_special_classes = 'right-8 top-8 rounded-xl';

---

The question is now:

Can I place such a string somewhere inside the theme that I can access it like:

```theme.my_special_classes```

or something equal.

Then we just have to adjust the tailwindcss.config.js as for colors also for such styles.

---

Hopefully there are some guides about what is the best approach in handling.

Thanks
Tom

**Comment by TomFreudenberg** [comment]
By the way, I know to create a CSS class with some `@apply` styles but I wonder if I can just do it by a simple string inside the theme instead working seperately on the css files.

**Comment by wongjn** [comment]
I'm not sure I fully understanding your requirements, but you could consider storing the `my_special_classes` variable in some `.js` file (since it seems like you're working with React or similar) and then you can `import` that variable wherever you need its value.

**Comment by TomFreudenberg** [comment]
Hey @wongjn thanks for coming around.

I found something like CVA and that pretty seems to fit what I am looking for.

More the question now to the experts: Is the cva approach "against" the tailwindcss idea or does both fit? CVA is by default also used inside Shadcn/UI (Buttons)

https://cva.style/

and an idea to pack that as a plugin for tailwind

https://www.youtube.com/watch?v=QJlTWj30krw

Thanks for some feedback about that
